plugins/amd-gpu/fu-amd-gpu-atom-firmware.c

```diff
@@ -0,0 +1,424 @@
+				XbBuilderNode *bn)
+{
+	FuAmdGpuAtomFirmware *self = FU_AMD_GPU_ATOM_FIRMWARE(firmware);
+
+	FU_FIRMWARE_CLASS(fu_amd_gpu_atom_firmware_parent_class)->export(firmware, flags, bn);
+
+	fu_xmlb_builder_insert_kv(bn, "part_number", self->part_number);
+	fu_xmlb_builder_insert_kv(bn, "asic", self->asic);
+	fu_xmlb_builder_insert_kv(bn, "pci_type", self->pci_type);
+	fu_xmlb_builder_insert_kv(bn, "memory_type", self->memory_type);
+	fu_xmlb_builder_insert_kv(bn, "bios_date", self->bios_date);
+	fu_xmlb_builder_insert_kv(bn, "model_name", self->model_name);
+	fu_xmlb_builder_insert_kv(bn, "config_filename", self->config_filename);
+}
+
+static gboolean
+fu_amd_gpu_atom_firmware_check_magic(FuFirmware *firmware, GBytes *fw, gsize offset, GError **error)
+{
+	gsize bufsz = 0;
+	const guint8 *buf;
+	g_autoptr(GByteArray) atom = NULL;
+
+	buf = g_bytes_get_data(fw, &bufsz);
+
+	if (!fu_struct_atom_image_validate(buf, bufsz, offset, error))
+		return FALSE;
+
+	atom = fu_struct_atom_image_parse(buf, bufsz, offset, error);
+	if (atom == NULL)
+		return FALSE;
+
+	return fu_struct_atom_rom21_header_validate(buf,
+						    bufsz,
+						    fu_struct_atom_image_get_rom_loc(atom),
+						    error);
+}
+
+const gchar *
+fu_amd_gpu_atom_get_vbios_pn(FuFirmware *firmware)
+{
+	FuAmdGpuAtomFirmware *self = FU_AMD_GPU_ATOM_FIRMWARE(firmware);
+	return self->part_number;
+}
+
+static gboolean
+fu_amd_gpu_atom_parse_vbios_version(FuAmdGpuAtomFirmware *self,
+				    const guint8 *buf,
+				    gsize bufsz,
+				    GError **error)
+{
+	guint8 vbios_version[BIOS_STRING_LENGTH];
+	gsize offset, base;
+
+	base = fu_firmware_get_addr(FU_FIRMWARE(self));
+	if (!fu_memmem_safe(buf + base,
+			    bufsz - base,
+			    (const guint8 *)BIOS_VERSION_PREFIX,
+			    sizeof(BIOS_VERSION_PREFIX) - 1,
+			    &offset,
+			    error)) {
+		g_prefix_error(error, "failed to find anchor: ");
+		return FALSE;
+	}
+
+	/* skip anchor */
+	offset += sizeof(BIOS_VERSION_PREFIX) - 1 + base;
+
+	if (!fu_memcpy_safe(vbios_version,
+			    sizeof(vbios_version),
+			    0x0,
+			    buf,
+			    bufsz,
+			    offset,
+			    BIOS_STRING_LENGTH,
+			    error)) {
+		g_prefix_error(error, "failed to get vbios version: ");
+		return FALSE;
+	}
+
+	fu_firmware_set_version(FU_FIRMWARE(self), (gchar *)vbios_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_amd_gpu_atom_parse_vbios_date(FuAmdGpuAtomFirmware *self, GByteArray *atom_image, GError **error)
+{
+	g_autoptr(GByteArray) st = NULL;
+	gchar vbios_date[17] = "20XX/XX/XX XX:XX";
+	guint16 val;
+
+	st = fu_struct_atom_image_get_vbios_date(atom_image);
+	if (st == NULL)
+		return FALSE;
+
+	/* this is to be able to reflect the exact same date format used by atom_get_vbios_date() */
+	val = fu_struct_vbios_date_get_year(st);
+	if (!fu_memcpy_safe((guint8 *)vbios_date,
+			    sizeof(vbios_date),
+			    2,
+			    (guint8 *)&val,
+			    sizeof(val),
+			    0,
+			    sizeof(val),
+			    error))
+		return FALSE;
```

Comment:
wouldn't all this be much simpler and safer if it was just `g_strdup_printf("20%02X/%02X/%02X %02X:%02X", fu_struct_vbios_date_get_year(st), ....)`